sanic_routing/tree.py

```diff
@@ -57,186 +88,230 @@ def display(self) -> None:
-                        [
-                            Line(
-                                f"{if_stmt} method in {route.methods}:",
-                                return_indent,
-                            ),
-                            Line(f"route_idx = {i}", return_indent + 1),
-                        ]
+                return_indent += 1
+
+            for group in sorted(self.groups, key=self._group_sorting):
+                group_bump = 0
+
+                # If the route had some requirements, let's make sure we check
+                # them in the source
+                if group.requirements:
+                    route_idx = "route_idx"
+                    self._inject_requirements(
+                        location, return_indent + group_bump, group
+                    )
+
+                # This is for any inline regex routes. It sould not include,
+                # path or path-like routes.
+                if group.regex:
+                    self._inject_regex(
+                        location, return_indent + group_bump, group
                     )
-                location.extend(
-                    [
-                        Line("else:", return_indent),
-                        Line("raise NoMethod", return_indent + 1),
-                    ]
+                    group_bump += 1
+
+                # Since routes are grouped, we need to know which to select
+                # Inside the compiled source, we keep track so we know which
+                # handler to assign this to
+                if route_idx == 0 and len(group.routes) > 1:
+                    route_idx = "route_idx"
+                    self._inject_method_check(
+                        location, return_indent + group_bump, group
+                    )
+
+                # The return.kingdom
+                self._inject_return(
+                    location, return_indent + group_bump, route_idx, group
                 )
-            routes = "regex_routes" if self.group.regex else "dynamic_routes"
-            route_return = (
-                "" if self.group.router.stacking else f"[{route_idx}]"
-            )
+
+        return src, delayed, final
+
+    def add_child(self, child: "Node") -> None:
+        self._children[child.part] = child
+
+    def _inject_param_check(self, location, indent, idx):
+        """
+        Try and cast relevant path segments.
+        """
+        unquote_start = "unquote(" if self.unquote else ""
+        unquote_end = ")" if self.unquote else ""
+        lines = [
+            Line("try:", indent),
+            Line(
+                f"basket['__matches__'][{idx}] = "
+                f"{unquote_start}{self.param.cast.__name__}(parts[{idx}])"
+                f"{unquote_end}",
+                indent + 1,
+            ),
+            Line("except ValueError:", indent),
+            Line("pass", indent + 1),
+            Line("else:", indent),
+        ]
+        self.base_indent += 1
```

Comment:
I gather that the `base_indent` increment is to account for the `else:` block but `indent` is also incremented where this function is called. Is this intentional?

Reply:
Possible optimization: do unquoting on the whole URL, or at least only once per segment instead of separately for each test (benchmarked 250 ns per unquote call, not much but about 10 times more than str eq comparison).

Reply:
> I gather that the `base_indent` increment is to account for the `else:` block but `indent` is also incremented where this function is called. Is this intentional?

Yes. This is one of the rare instances where we increment the `base_indent` because of its impact on subsequent nodes.



> Possible optimization: do unquoting on the whole URL, or at least only once per segment instead of separately for each test (benchmarked 250 ns per unquote call, not much but about 10 times more than str eq comparison).

:+1:  Thanks for testing that.

Reply:
@Tronic 

![image](https://user-images.githubusercontent.com/166269/120119143-bde7a300-c19e-11eb-8918-2b12cc40bbb2.png)

Currently the code does unquoting outside of the casting. There are two possible approaches I can think of: do it early or do it late.

```
parts = tuple(map(unquote, parts))  # do once for node if needed
try:
    ...
```

or

```
try:
    ...
...
else:
    basket['__matches__'][1] = unquote(basket['__matches__'][1])
```

The second approach would follow the existing logic in doing the casting before unquoting. I am thinking maybe to go with the first approach, but it would be a change. Can you think of any reasons this would be a problem. The more I think about it, the more I think that you would want it to be unquoted, especially in the context of adding custom casting operations. (See #30)

Reply:
So I was thinking that we could do it once upfront for each test. But, that will not work since some routes might be quoted, and some not. Therefore, I think the best place for this to reduce unnecessary options would be to apply inside the `else` block.

Reply:
![image](https://user-images.githubusercontent.com/166269/120119688-cdb4b680-c1a1-11eb-969b-a789b08201c2.png)